api/app/controllers/api/v1/offers_controller.rb

```diff
@@ -1,31 +1,396 @@
+        position = Position.find_by(position: offer["course_id"])
+        applicant = Applicant.find_by(utorid: offer["utorid"])
+        if position && applicant
+          ident = {position_id: position[:id], applicant_id: applicant[:id]}
+          exists = "offer with position #{offer[:position]} for applicant #{offer[:utorid]} already exists"
+          data = {
+            position_id: position[:id],
+            applicant_id: applicant[:id],
+            hours: offer["hours"],
+            session: offer["session"],
+            year: offer["year"],
+          }
+          begin
+            insertion_helper(Offer, data, ident, exists)
+          rescue
+            exceptions.push("Error: Unknown error when creating Position '#{offer["course_id"]}' for '#{offer["utorid"]}' for '#{data[:hours]}' hours.")
+          end
+        else
+          if position
+            exceptions.push("Error: Applicant #{offer["utorid"]} is invalid.")
+          else
+            exceptions.push("Error: either Position #{offer["course_id"]} or Applicant #{offer["utorid"]} is invalid.")
+          end
+        end
+      end
+
+      if exceptions.length == offers.length
+        status = {success: false, errors: true, message: exceptions}
+      elsif exceptions.length > 0
+        status = {success: true, errors: true, message: exceptions}
+      else
+        status = {success: true, errors: false, message: ["Offers import was successful."]}
+      end
+
+      if status[:success]
+        render json: {errors: status[:errors], message: status[:message]}
+      else
+        render status: 404, json: {message: status[:message], errors: status[:errors]}
+      end
+    end
+
+    def update
+      if params[:id] == "batch-update"
+        if params[:offers]
+          params[:offers].each do |id|
+            offer = Offer.find(id)
+            offer.update_attributes!(offer_params)
+          end
+        else
+          render status: 404, json: {message: "Error: No offers given."}
+        end
+      else
+        offer = Offer.find(params[:id])
+        offer.update_attributes!(offer_params)
+      end
+    end
+
+    def send_contracts
+      begin
+        params[:offers].each do |id|
+          offer = Offer.find(id)
+            if ENV['RAILS_ENV'] != 'test'
+              offer.update_attributes!(link: "/pb/#{offer[:id]}")
+              CpMailer.contract_email(offer.format,"#{ENV["domain"]}#{offer[:link]}").deliver_now!
+            end
+            offer.update_attributes!({status: "Pending", send_date: DateTime.now.to_s})
+        end
+        render status: 200, json: {message: "You've successfully sent out all the contracts."}
+      rescue Errno::ECONNREFUSED
+        puts "rejected"
+        render status: 404, json: {message: "Connection Refused."}
+      end
+    end
+
+    def batch_email_nags
+      begin
+        params[:offers].each do |id|
+          offer = Offer.find(id)
+          offer.increment!(:nag_count, 1)
```

Comment:
Better to increment the counter only after the email has been sent successfully.